This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = '8LH5zy6VseHgv4Txh-ux'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests
import collections
import json
import numpy as np

In [3]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url_0 = "https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2020-01-13&end_date=2020-01-13&api_key=" + API_KEY
response = requests.get(url_0)
#check of response - 200 means that the request is successful
response.status_code

200

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure.
json_0 = response.json()
print(json_0)

{'dataset_data': {'limit': None, 'transform': None, 'column_index': None, 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'start_date': '2020-01-13', 'end_date': '2020-01-13', 'frequency': 'daily', 'data': [['2020-01-13', 118.0, 118.9, 117.0, 117.2, None, 84067.0, 9891618.0, None, None, None]], 'collapse': None, 'order': None}}


In [5]:
#Further Inspection of Levels in the Json file
#Level 1 - We find the keys
json_0.keys()

dict_keys(['dataset_data'])

In [6]:
#level 2 - We search the second key level
json_0['dataset_data'].keys()

dict_keys(['limit', 'transform', 'column_index', 'column_names', 'start_date', 'end_date', 'frequency', 'data', 'collapse', 'order'])

In [7]:
#Level 3 - We search the third key level
json_0['dataset_data']['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [8]:
#Level 3 - We search the third key level
json_0['dataset_data']['data']

[['2020-01-13',
  118.0,
  118.9,
  117.0,
  117.2,
  None,
  84067.0,
  9891618.0,
  None,
  None,
  None]]

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

# 1 Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD)#

In [9]:
#Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
#We define the stock market and the stock to retrieve
db1 = 'FSE'
db2 = 'AFX_X'
#Definition of the url of the request
url_2017 = 'https://www.quandl.com/api/v3/datasets/' + db1 + '/' + db2 + '/data'
#Definition of parameters of the request
params = dict(key=API_KEY,
              name = 'AFX_X - 2017 data',
              return_format = '.json',
              start_date = '2017-01-01',
              end_date='2017-12-31')
#request of data to the API
r1 = requests.get(url_2017, params=params)
#check of the response - 200 means data was correctly retrieved
r1.status_code

200

**The request of data is sucessfull as status code is 200**

# 2 Convert the returned JSON object into a Python dictionary

In [10]:
#Convert the returned JSON object into a Python dictionary
json_2017 = r1.json()
type(json_2017)

dict

# 3 Calculate what the highest and lowest opening prices were for the stock in this period.

In [11]:
#We need an index of "Open" to get a list from the Dictionary
json_2017['dataset_data']['column_names'].index('Open')

1

In [12]:
#Make a list of the Open price from Dictionary
d2017 = json_2017['dataset_data']['data']
list_open = []
for i in d2017:
    list_open.append(i[1])

In [13]:
#Calculate max open price, ignoring the None values
max([i for i in list_open if i is not None])

53.11

In [14]:
#Calculate min open price, ignoring None values
min([i for i in list_open if i is not None])

34.0

**The highest opening price in 2017 is 53.11 euros, the lowest opening price is 34 euros**

# 4 What was the largest change in any one day (based on High and Low price)?

In [15]:
#Get index of High to make a list from Dictionary
json_2017['dataset_data']['column_names'].index('High')

2

In [16]:
#Get index of Low to make a list from Dictionary
json_2017['dataset_data']['column_names'].index('Low')

3

In [17]:
#Create 2 lists, one with High values and another with Low values
d2017 = json_2017['dataset_data']['data']
list_open_high = []
list_open_low = []
for i in d2017:
    list_open_high.append(i[2])
    list_open_low.append(i[3])

In [18]:
#Create a function to subtract the Low values to the High values
def subtract(L1,L2): 
   return [x1-x2 for x1,x2 in zip(L1,L2)]

In [19]:
#Output the max value of the subtacted list
max(subtract(list_open_high,list_open_low))

2.8100000000000023

**The highest change in price within any day is 2.81 euros**

# 5 What was the largest change between any two days (based on Closing Price)?

In [20]:
#Identify index of Close value to use it later to create a list from Dictionary
json_2017['dataset_data']['column_names'].index('Close')

4

In [21]:
#Create a list of Close value for Dictionary
list_close = []
for i in d2017:
    list_close.append(i[4])

In [22]:
#Create a list of max positive differences in Close value: 2.55 euros is the max positive difference
max([y - x for x,y in zip(list_close,list_close[1:])])

2.559999999999995

In [23]:
#Create a list of max negative differences in Close value: -1.71 euros is the max neg difference
min([y - x for x,y in zip(list_close,list_close[1:])])

-1.7199999999999989

**The largest change between any 2 days is 2.55 euros**

# 6 What was the average daily trading volume during this year?

In [24]:
#Identify index of Trading volume to use it later to create a list from Dictionary
json_2017['dataset_data']['column_names'].index('Traded Volume')

6

In [25]:
#Create a list of Trading volume value for Dictionary
list_volume = []
for i in d2017:
    list_volume.append(i[6])

In [26]:
#Calculate mean by adding all values of list Trading volume / nº of observations: 89.124
sum(list_volume) / len(list_volume)

89124.33725490196

**The mean of the Trading volume in 2017 is 89.124**

# 7 (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [27]:
#We create a function to calculate the median value of a list
def median(lst):
    n = len(lst)
    s = sorted(lst)
    return (sum(s[n//2-1:n//2+1])/2.0, s[n//2])[n % 2] if n else None

In [28]:
#We calculate the median Trading volume of 2017: 76.286
median(list_volume)

76286.0

**The median of the Trading volume in 2017 is 76.286**